bottlecap/src/traces/span_pointers.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use serde_json::{json, Value};
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+
+const SPAN_POINTER_HASH_LENGTH: usize = 32;
+
+#[derive(Clone)]
+pub struct SpanPointer {
+    pub hash: String,
+    pub kind: String,
+}
+
+/// Returns the first 32 characters of the SHA-256 hash of the components joined by a '|'.
+/// Used by span pointers to uniquely & deterministically identify an `S3` or `DynamoDB` stream.
+/// <https://github.com/DataDog/dd-span-pointer-rules/blob/main/README.md#General%20Hashing%20Rules>
+#[must_use]
+pub fn generate_span_pointer_hash(components: &[&str]) -> String {
+    let mut hasher = Sha256::new();
+    hasher.update(components.join("|").as_bytes());
+    let result = hasher.finalize();
+    hex::encode(result)[..SPAN_POINTER_HASH_LENGTH].to_string()
+}
+
+pub fn attach_span_pointers_to_meta<S: ::std::hash::BuildHasher>(
+    meta: &mut HashMap<String, String, S>,
+    span_pointers: &[SpanPointer],
+) {
+    if span_pointers.is_empty() {
+        return;
+    }
+
+    let span_links: Vec<Value> = span_pointers
+        .iter()
+        .map(|sp| {
+            json!({
+                "attributes": {
+                    "link.kind": "span-pointer",
+                    "ptr.dir": "u",
+                    "ptr.hash": sp.hash,
+                    "ptr.kind": sp.kind,
+                },
+                "span_id": "0",
+                "trace_id": "0"
```

Comment:
Make sure to use the `trace-protobuf` implementation of `libdatadog` for the `SpanLink` structure instead of the JSON you have here